src/main/java/org/opensearch/ad/ml/EntityColdStarter.java

```diff
@@ -476,27 +456,176 @@ private void getEntityColdStartData(String detectorId, Entity entity, ActionList
+                getFeatures(
+                    listener,
+                    round + 1,
+                    currentRoundColdStartData,
+                    detector,
+                    entity,
+                    stride,
+                    numberOfSamples,
+                    startTimeMs,
+                    lastSampleStartTime
+                );
+            }
+        }, listener::onFailure);
+
+        try {
+            searchFeatureDao
+                .getColdStartSamplesForPeriods(
+                    detector,
+                    sampleRanges,
+                    entity,
+                    // Accept empty bucket.
+                    // 0, as returned by the engine should constitute a valid answer, “null” is a missing answer — it may be that 0
+                    // is meaningless in some case, but 0 is also meaningful in some cases. It may be that the query defining the
+                    // metric is ill-formed, but that cannot be solved by cold-start strategy of the AD plugin — if we attempt to do
+                    // that, we will have issues with legitimate interpretations of 0.
+                    true,
+                    new ThreadedActionListener<>(logger, threadPool, AnomalyDetectorPlugin.AD_THREAD_POOL_NAME, getFeaturelistener, false)
+                );
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private int calculateColdStartDataSize(List<double[][]> coldStartData) {
+        int size = 0;
+        for (int i = 0; i < coldStartData.size(); i++) {
+            size += coldStartData.get(i).length;
+        }
+        return size;
+    }
+
+    /**
+     * Select strideLength and numberOfSamples, where stride is the number of intervals
+     * between two samples and trainSamples is training samples to fetch.
+     *
+     * Algorithm:
+     *
+     * delta is the length of the detector interval in minutes.
+     *
+     * 1. Suppose delta ≤ 30 and divides 60. Then set numberOfSamples = ceil ( (shingleSize + 32)/ 24 )*24
+     * and strideLength = 60/delta. Note that if there is enough data — we may have lot more than shingleSize+32
+     * points — which is only good. This step tries to match data with hourly pattern.
+     * 2. Set numberOfSamples = (shingleSize + 32) and strideLength = 1. This should be an uncommon case,
+     * but if someone wants 23 minutes interval — and the system permits lets give it to them. Note the
+     * smallest delta that does not divide 60 is 7 which is quite large to wait for one data point.
+     * @return the chosen strideLength and numberOfSamples
+     */
+    private Pair<Integer, Integer> selectRangeParam(AnomalyDetector detector) {
+        long delta = detector.getDetectorIntervalInMinutes();
+        int shingleSize = detector.getShingleSize();
+        int strideLength = defaulStrideLength;
+        int numberOfSamples = defaultNumberOfSamples;
+        if (delta <= 30 && 60 % delta == 0) {
```

Comment:
if  (delta == 1) then the % operator fails I think (even though 1 divides 60) ...

Reply:
jshell> 60 % 2
$1 ==> 0
jshell> 60 % 1
$2 ==> 0
jshell> 60 % 7
$3 ==> 4

It seems 1 works.

Reply:
ok